include/xtensor/xshape.hpp

```diff
@@ -0,0 +1,687 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht    *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_XSHAPE_HPP
+#define XTENSOR_XSHAPE_HPP
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <cstdlib>
+#include <cstring>
+#include <initializer_list>
+#include <iterator>
+#include <memory>
+
+#include "xexception.hpp"
+#include "xstorage.hpp"
+
+#if (__GNUC__ || __clang__)
+    #define likely(x)       __builtin_expect((x), 1)
+    #define unlikely(x)     __builtin_expect((x), 0)
+#else
+    #define likely(x)       x
+    #define unlikely(x)     x
+#endif
+
+#ifndef XSHAPE_ALIGNMENT
+#ifdef XTENSOR_USE_XSIMD
+#include "xsimd/xsimd.hpp"
+#define XSHAPE_ALIGNMENT XSIMD_DEFAULT_ALIGNMENT
+#else
+#define XSHAPE_ALIGNMENT T
+#endif
+#endif
+
+namespace xt
+{
+    template <class T, std::size_t N, class A = std::allocator<T>, bool INIT = true>
+    class small_vector;
+
+    template <class T>
+    using dynamic_shape = small_vector<T, 4>;
+
+    template <class T, std::size_t N>
+    using static_shape = std::array<T, N>;
+
+    template <std::size_t... X>
+    class fixed_shape;
+
+    template <class T, std::size_t N, class A, bool INIT>
+    class small_vector
```

Comment:
Global comment: the `small_vector` has a more general purpose than xtensor shapes, so it should be in a dedicated file (that would be moved to the `xtl` eventually). 

Besides I don't find the name explicit enough, since it can also be used as a large vector.

Reply:
I'll move it. stack_vector? stackbuffer_vector? fast_vector? svector? tector?

Reply:
Actually can you move it in `xstorage.hpp` ?
For the name, ovector (for optimized_vector) or fvector (for fast_vector) so it is consistent with the `uvector`, what do you think?

Reply:
how about `svector` for `stack, smart & small` :)

Reply:
So be it!